File: jsyd_102/src/SSCScore.java
import java.io.*;
import java.util.*;

public class SSCScore extends SSScore
{
  public String scoreFileSpec = "";

  SSCScore(ModList mList)
  {
    super(MT_CScore, mList);
  }

  static  double[] lastP = new double[16];

  double GrabParam(String p, int pNbr)
  {
    double v;
    if (p.equals("+") && pNbr == 2)
      v = lastP[2]+lastP[3];
    else if (p.equals("."))
      v = lastP[pNbr];
    else
      v = Double.parseDouble(p);
    lastP[pNbr] = v;
    return v;
  }

  void GenerateScore(SSModule callingMod)
  {
    // unsigned char *buffer;
    // ScoreSectionPtr sp;
    // NoteEventPtr  np;
    // char    *p;

    // Open the file
    BufferedReader ar = null;
    try {
      ar = new BufferedReader(new FileReader(scoreFileSpec));
    }
    catch (IOException e)
    {
      System.out.println("Couldn't open " + scoreFileSpec + " sydspec=" + PatchOwner.sydSpec);
      // !!! Try to open using path in global sydSpec
      int     li = PatchOwner.sydSpec.lastIndexOf('/');
      if (li == -1)
        li = PatchOwner.sydSpec.lastIndexOf('\\');
      if (li >= 0) {
        scoreFileSpec = PatchOwner.sydSpec.substring(0,li+1) + scoreFileSpec;
        System.out.println("Can't open score, trying " + scoreFileSpec);
        try {
          ar = new BufferedReader(new FileReader(scoreFileSpec));
        }
        catch (IOException e2) {
          System.out.println("Error opening score " + scoreFileSpec + ": " + e2.toString());
          parList.itsOwner.AbortSynthesis();
          return;
        }
      }
      else {
        System.out.println("Error opening score (retry path?): " + scoreFileSpec + ": " + e.toString());
        parList.itsOwner.AbortSynthesis();
        return;
      }
    }

    if (ar == null) {
      System.out.println("Can't open score file");
      parList.itsOwner.AbortSynthesis();
      return;
    }
    System.out.println("Resetting");
    // Reset vars
    sections.clear();
    AddSection();
    ScoreSection sp = (ScoreSection) sections.firstElement();

    System.out.println("Initting");
    sp.sectionStart = 0.0;
    sp.tempoScale = 1.0;
    sp.tempoStart = 0.0;
    sp.timeStart = 0.0;

    String  p;

    try {

    while ((p = ar.readLine()) != null &&
           parList.itsOwner.windowState == WS_Synthesize)
    {
      int pcIdx = 0;
      int pLen = p.length();

      while (pcIdx < pLen && Character.isWhitespace(p.charAt(pcIdx)))
        ++pcIdx;

      if (pcIdx >= pLen || p.charAt(pcIdx) == ';' || p.charAt(pcIdx) == '#')
      {
        continue;
      }
      StringTokenizer st = new StringTokenizer(p.substring(pcIdx));
      String lab = st.nextToken();
      if (lab.equals("t")) // tempo
      {
        double p1 = GrabParam(st.nextToken(),1);
        double p2 = GrabParam(st.nextToken(),2);

        sp.timeStart = (p1 - sp.tempoStart) * sp.tempoScale + sp.timeStart;
        sp.tempoStart = p1;
        sp.tempoScale = 60 / p2;
      }
      else if (lab.equals("f")) // table  (f0 is used to create silence)
      {
        // !!! Add wave table event NEF_Function
      }
      else if (lab.equals("i")) // note event
      {
                    // !!!
                    // '<' is for ramping (mark 'begin ramp')
                    //  '>' treat as for ramping as well
        double p1 = GrabParam(st.nextToken(),1);
        double p2 = GrabParam(st.nextToken(),2);
        double p3 = GrabParam(st.nextToken(),3);
        double[] optP = new double[16];
        int     nbrOpts = 0;
        while (st.hasMoreTokens())
        {
           optP[nbrOpts++] = GrabParam(st.nextToken(),4+nbrOpts);
        }
        NoteEvent np = AddNoteEvent(sp,0,(int) p1,p2,p3,nbrOpts);
        System.arraycopy(optP,0,np.op,0,nbrOpts);
      }
      else if (lab.equals("s")) // section end
      {
        ScoreSection lastS = sp;
        AddSection();
        sp = (ScoreSection) sections.lastElement();

        sp.sectionStart = lastS.sectionStart+lastS.sectionLength;
        sp.tempoScale = 1.0;
        sp.tempoStart = 0.0;
        sp.timeStart = 0.0;
      }
      else if (p.charAt(pcIdx) == 'e') // end of score
      {
        break;
      }
      else {
        System.out.println("Invalid line in score: " + p);
        parList.itsOwner.AbortSynthesis();
      }
    }
    ar.close();
    }
    catch (IOException e)
    {
        System.out.println("Exception reading C Score: " + e.toString());
        parList.itsOwner.AbortSynthesis();
    }

    if (PatchOwner.gDurationOverride) {
      // System.out.println("Override duration?");
      double scoreDuration = (sp.sectionStart + sp.sectionLength) * sp.tempoScale;
      SSOutput outMod = (SSOutput) (parList.itsOwner.outMod);
      if (outMod != null && scoreDuration > outMod.sampleDuration)
      {
        System.out.println("Overriding duration " + outMod.sampleDuration + " -> " + scoreDuration);
        outMod.sampleDuration = scoreDuration;
        parList.itsOwner.mainInst.sampleDuration = scoreDuration;
      }
    }
  }

  void Save(PrintWriter ar) throws IOException
  {
    super.Save(ar);
    ar.println("CSCO " + scoreFileSpec);
  }

  void Load(BufferedReader ar) throws IOException
  {
    scoreFileSpec = getFirstToken(ar,"CSCO");
  }

  void Copy(SSModule ss)
  {
    SSCScore sc = (SSCScore) ss;
    super.Copy(ss);
    this.scoreFileSpec = sc.scoreFileSpec;
  }


}

File: csyd_102/src/sscscore.cpp
// sscscore.cp

#include "ss.h"
#include "sscscore.h"
#include "patchowner.h"
#include "ssoutput.h"

SSCScore::SSCScore(ModList * mList, short h, short v) : SSScore(MT_CScore, mList, h, v)
{
	scoreFileSpec.name[0] = 0;
	// scoreFileSpec.vRefNum = 0;
	// scoreFileSpec.parID = 0;
}

char *SSCScore::GrabParam(char *p, int pNbr, ssfloat *pStore, Boolean *gotIt)
{
	static	ssfloat lastP[16];

	while (*p && (*p == ' ' || *p == '\t'))
		++p;
	if (*p == '\r' || *p == '\n') {
		*gotIt = false;
		return p;
	}
	if (*p == '+' && pNbr == 2) {
		*pStore = lastP[2]+lastP[3];
		++p;
	}
	else if (*p == '.' && !isdigit(p[1])) {
		*pStore = lastP[pNbr];
		++p;
	}
	else {
		*pStore = atof((char *) p);
		while (*p && !isspace(*p))
			++p;
	}
	lastP[pNbr] = *pStore;
	*gotIt = true;
	return p;
}

void SSCScore::GenerateScore(SSModule *callingMod)
{
	unsigned char *buffer;
	ScoreSectionPtr	sp;
	NoteEventPtr	np;
	char		*p;

	// Open the file
	buffer = SuckFilefromSpec(&scoreFileSpec);	// Locally implemented
	if (buffer == NULL) {
		ErrorMessage("Can't open score file");
		parList->itsOwner->AbortSynthesis();
		return;
	}
	
	p = (char *) buffer;
	// Reset vars
	nbrSections = 0;
	sections = NULL;
	AddSection();

	sp = sections;
	sp->sectionStart = 0.0;
	sp->tempoScale = 1.0;
	sp->tempoStart = 0.0;
	sp->timeStart = 0.0;

	while (*p && parList->itsOwner->windowState == WS_Synthesize) {
		if (isspace(*p))
			++p;
		else if (*p == ';' || *p == '#')
		{
			while (*p && *p != '\r' && *p != '\n')
				++p;
		}
		else if (*p == 't')	// tempo
		{
			ssfloat	p1,p2;
			Boolean	gotIt;
			++p;
			p = GrabParam(p,1,&p1,&gotIt);
			p = GrabParam(p,2,&p2,&gotIt);

			sp->timeStart = (p1 - sp->tempoStart) * sp->tempoScale + sp->timeStart;
			sp->tempoStart = p1;
			sp->tempoScale = 60 / p2;
			// np = AddNoteEvent(sp,NEF_Tempo,((int) p1)-1,p2,0,0);
			while (*p && *p != '\r' && *p != '\n')
				++p;
		}
		else if (*p == 'f')	// table  (f0 is used to create silence)
		{
			// !!! Add wave table event NEF_Function
			while (*p && *p != '\r' && *p != '\n')
				++p;
		}
		else if (*p == 'i')	// note event
		{
	//  If 'i', parse instrument  '+' in p2 = p2+p3   '.' = previous field  
									// '<' is for ramping (mark 'begin ramp')
									//  '>' treat as for ramping as well
			ssfloat	p1,p2,p3;
			ssfloat optP[16];
			int	nbrOpts = 0;
			int n,i;
			Boolean	gotIt;
			++p;
			p = GrabParam(p,1,&p1,&gotIt);
			p = GrabParam(p,2,&p2,&gotIt);
			p = GrabParam(p,3,&p3,&gotIt);
			n = 4;
			do {
				p = GrabParam(p,n,&optP[nbrOpts],&gotIt);
				++n;
				if (gotIt)
					++nbrOpts;
			} while (gotIt);
			np = AddNoteEvent(sp,0,(int) p1,p2,p3,nbrOpts);
			for (i = 0; i < nbrOpts; ++i)
				np->op[i] = optP[i];
		}
		else if (*p == 's')	// section end
		{
			ScoreSectionPtr	lastS = sp;
			AddSection();
			sp = sections + (nbrSections-1);
			sp->sectionStart = lastS->sectionStart+lastS->sectionLength;
			sp->tempoScale = 1.0;
			sp->tempoStart = 0.0;
			sp->timeStart = 0.0;
			while (*p && *p != '\r' && *p != '\n')
				++p;
		}
		else if (*p == 'e')	// end of score
		{			
			break;
		}
		else {
			ErrorMessage("Invalid line in score: %.20s",p);
			parList->itsOwner->AbortSynthesis();
		}
	}
  if (gDurationOverride) {
    ssfloat scoreDuration = (sp->sectionStart + sp->sectionLength)*sp->tempoScale;
    SSOutput *outMod = (SSOutput *) (parList->itsOwner->outMod);
    if (outMod != NULL && scoreDuration > outMod->sampleDuration)
    {
      LogMessage("Overriding duration %g -> %g\n", outMod->sampleDuration, scoreDuration);
      outMod->sampleDuration = scoreDuration;
      parList->itsOwner->mainInst->sampleDuration = scoreDuration;
    }
  }
	MyDisposePtr(buffer);
	buffer = NULL;
}



/*
void SSCScore::Save(FILE* ar)
{

	SSScore::Save(ar);

#if macintosh
	char	tempS[64];

	BlockMove(scoreFileSpec.name,tempS,scoreFileSpec.name[0]+1);
	PtoCstr((StringPtr) tempS);
	WriteFileLine(ar,"CSCO %s %d %ld\r",
					tempS,
					(int) scoreFileSpec.vRefNum, 
					scoreFileSpec.parID);
#else
	WriteFileLine(ar,"CSCO %s 0 0\r",scoreFileSpec.name);
#endif
}
*/

void SSCScore::Load(FILE* ar) 
{
	char				*p,tbuf[kMaxLineLength];
	// int					vRefNum;
	// long				parID;
	char				tempS[kMaxFilenameLength];

	p = parList->GetNextInputLine(ar,"CSCO",tbuf);
  // sscanf(p, "CSCO %s %d %ld",tempS,&vRefNum,&parID);
  sscanf(p, "CSCO %s",tempS);
	// scoreFileSpec.vRefNum = vRefNum;
	// scoreFileSpec.parID = parID;
	strcpy(scoreFileSpec.name,tempS);
}

/*
char* SSCScore::GetLabel()
{
	return scoreFileSpec.name;
}
*/

void SSCScore::Copy(SSModule *ss)
{
	SSCScore *sc = (SSCScore *) ss;
	SSScore::Copy(ss);
	this->scoreFileSpec = sc->scoreFileSpec;
}
